쥬스 메이커 [STEP 1] Hugh, yeton

Hugh-github opened the pull request:
안녕하세요😆 제이슨 @ehgud0670 
step1 PR보냅니다.

## step1 기능 구현
- 구조체(JuiceMaker),클래스(FruitStore),프로토콜(BindingOptional),enum(Fruits,Menu,JuiceMakerError) 구현
- 구조체(JuiceMaker) 
    - 재고가 충분한지 확인하는 메서드(checkStock)
    - 주스를 만드는 메서드(make)
    - 딕셔너리 타입의 Menu.recipe에서 key값만 분리해오는 메서드(classifyKey)
    - 프로토콜을 채택해서 옵셔널 바인딩해주는 메서드(unwrapOptional) 
- 클래스(FruitStore)
    - 재고를 감소시켜주는 메서드(decreaseStock)
    - 재고를 채워주는 메서드(fillStock)
    - 프로토콜을 채택해서 옵셔널 바인딩해주는 메서드(unwrapOptional) 
- enum(Menu)
    - 연산프로퍼티를 사용해서 주스의 레시피를 딕셔너리 타입으로 반환해줌 
- protocol(BindingOptional)
    - JuiceMaker, FruitStore에서 옵셔널 바인딩 해주는 부분이 필요해서 프로토콜을 따로 만들어줌

## 고민했던 것들
1. 옵셔널 바인딩 시 받아주는 상수명을 어떻게 써야할 지 고민이 됩니다.
```swift
    func unwrapOptional(_ stock: Int?) -> Int {
        guard let number = stock else { return 0 }
        return number
    }
```
위에서 number에 stock이 담기면 이라는 조건으로 옵셔널 바인딩을 해주고 싶었는데 네이밍을 어떤식으로 하면 좋을 지 감이 잘 안왔습니다..ㅜㅜ 🤯

2. 제이슨의 추천 영상에서 입력/처리/출력 부분을 다 객체로 분리해서 main함수에서 세 개의 객체가 서로 상호작용을 하도록 하는 방법이 좋다고 배웠습니다. 
     - 입력 처리를 분리하기 위해 노력을 해봤는데 생각이 잘 나지 않았습니다. 혹시 어떤 방향으로 작성해야 입력과 처리를 분리할 수 있는지 궁금합니다.

3. 재고를 확인해주는 메서드(checkStock)를 JuiceMaker에 구현해줄 지 FruitStore에 구현해줄 지 고민이 됐습니다. 
    - 저희는 재고를 확인하는 함수를 구조체(JuiceMaker)안에서 선언했는데 재고가 있는지 확인하는 기능이라고 생각했을 때 재고 속성을 가지고 있는 FruitStore에서 함수를 구현하는게 더 좋았을지 궁금합니다.

4. 들여쓰기를 최대 2번만 사용해야 하는 조건을 만족하기 위해서 코드가 길어지는 걸 방지하고자 옵셔널 바인딩을 해주는 메서드(unwrapOptional)를 만들고, 메서드 안에서 stock이 nil일 경우에 return 0을 해주었는데, 다른 방식으로는 어떻게 하면 좋을 지 고민이 됩니다. 
```swift
 func unwrapOptional(_ stock: Int?) -> Int {
        guard let number = stock else { return 0 }
        return number
    }
```

잘 부탁드립니다!! 😍

Hugh-github committed:
Create README.md

yeeton37 committed:
docs: README 파일 수정

구현할 타입, 프로퍼티, 메서드 정리

Hugh-github committed:
docs: 클래스 다이어그램 추가

Hugh-github committed:
feat: 열거형 타입 생성(Fruits, Menu)

열거형 타입(Menu) 내부에 연산 property를 이용하여 쥬스별 레시피 선언

yeeton37 committed:
feat: 딕셔너리 변수(stock), 딕셔너리 상수(orderQuantity) 선언

Hugh-github committed:
feat: JuiceMaker 속성 선언 및 속성 접근제어 수준 설정

yeeton37 committed:
feat: 재고 수량 변경하는 메서드(decreaseStock) 선언

메서드(decreaseStock): 주문 받았을 때 재고 수량을 감소시킴

yeeton37 committed:
refactor: 프로퍼티(recipe)의 접근제어 수정

FruitStore 내부에서 recipe를 이용하도록 하기 위해서 접근 제어 수준 변경

Hugh-github committed:
feat: 메서드(checkStock) 선언

재고가 쥬스 제작에 필요한 개수보다 많으면 감소시키고 true를 리턴해주고 적으면 재고를 채워주고 false를 리턴한다.

yeeton37 committed:
feat: 재고 채워주는 메서드(fillStock) 선언

- return 타입을 bool타입으로 선언할 경우에는 주스의 재료가 여러개일 때 모든 주스의 재료에 대해 확인이 불가능해서
return 타입을 제거해줌
- 메서드(fillStock): 재고가 부족할 경우 기존에 선언했던 orderQuantity의 값만큼 채워줌

Hugh-github committed:
feat: 메서드(hasFruit) 선언

쥬스를 만들기 위해 재료를 충분히 가지고 있는지 확인하는  메서드 선언

yeeton37 committed:
feat: 메서드(buy) 선언

재고를 확인해서 재고가 있을 경우 필요한 재고만큼 빼주고, 재고가 부족할 경우 채워주는 기능을 한다.

Hugh-github committed:
feat: 메서드(classifyKey) 선언

딕셔너리에서 키를 분리해 배열로 리턴

yeeton37 committed:
refactor: 메서드(fillStock)의 매개변수의 타입을 Fruits->[Fruits]로 수정 및 메서드(checkStock) 삭제

- 타입 수정과 관련된 인자값 수정
- 함수 중복으로 인한 메서드(checkStock) 삭제

Hugh-github committed:
feat: 열거형(ErrorCase) 선언

- 재고 부족시 에러를 던저주는 case 추가
- hasFruits에서 에러를 던지고 buy함수에서 에러를 처리하는 기능 구현

yeeton37 committed:
feat: 열거형(Menu) 파일 분리 및 프로토콜(BindingOptional) 선언

- JuiceMaker와 FruitStore 두 곳에서 옵셔널 바인딩을 해주기 위해 프로토콜(BindingOptional) 채택 후 구현

yeeton37 committed:
refactor: 타입명과 메서드명, 변수명, 매개변수명 수정

yeeton37 committed:
refactor: 메서드(checkStock, classifyKey)의 접근제어 수준 변경

yeeton37 committed:
docs: README.md 파일 수정

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
-
-// 과일 저장소 타입
-class FruitStore {
+class FruitStore: BindingOptional {
```
JuiceMaker 는 struct 이고 FruitStore 는 class 인데 이유가 있을까요?
레퍼런스 타입과 값타입에 대해서 잘 알고 계실까요?
둘 객체 모두 메소드를 갖고 있어 별 차이가 없어보인데 둘다 class 든 struct든 하나로 통일해야 할듯 합니다.
고민해주세요!

yeeton37 replied:
struct는 값타입이고 복사를 하기 때문에 struct를 쓰는 상황은 
1. 불변성이 필요한 데이터 타입
2. 공유될 필요가 없는 타입
3. 다른 타입으로부터 상속받거나, 자신을 상속할 필요가 없을때

class는 참조타입이고 복사가 아닌 참조를 하기 때문에 class를 쓰는 상황은
1. 앱의 여러 곳에서 사용하더라도 한 영역에서 적용한 수정이 다른 영역에서도 적용되야 하는 경우
2.  파일 관리나 네트워크 연결과 같은 작업을 다룰 때 
3. 상속이 필요할 때
라고 할 수 있을 것 같습니다! 

struct를 쓸 지 class를 쓸 지 고민이 되네요...

ehgud0670 replied:
네 잘 정리하셨네요! 
**보통 메소드가 있고 객체로서 협력할수 있는 역할이라면 class 를 사용하고,
모델 값 혹은 DTO(Data Transfer Object)의 역할이라면 struct 를 사용합니다.**

ps. 위의 JuiceMaker와 FruitStore은 메소드가 있어 class 의 성격을 띄지만 사실 지금 프로젝트에서는 별로 상관이 없어보여서 
그대로 두셔도 괜찮을 듯 합니다.

yeeton37 replied:
감사합니다. 이 개념을 봐도봐도 헷갈렸는데...ㅠㅠ 메소드,협력하는 역할, 모델 값 키워드로 보니 정리가 된 것 같습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/BindingOptional.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by hugh, yeton on 2022/04/27.
+//
+
+protocol BindingOptional {
```
프로토콜 타입을 써보는 것은 좋으나, 이미 스위프트에 `guard let` 이나 `If let` 로 언래핑하는 것을 제공해주고 아니면 `??` 을 사용해 nil일때 디폴트값을 쓸수 있도록 할수 있는데 굳이 있을 필요가 있을까요??
그리고 protocol은 여러 객체들에 대하여 범용적이면 좋은데(최소 2개 이상) 현재 stock을 가지고 있는 객체는 FruitStore 1개만 이므로 필요가 없어보이네요! JuiceMaker도 해당 프로토콜을 채택했지만 stock 을 가지고 있지 않으므로 부적절해보입니다.

Hugh-github replied:
guard let을 이용해 프로토콜을 제거하고 언래핑을 진행했습니다. 
`guard let number = stock[fruit]`

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
-class FruitStore {
+class FruitStore: BindingOptional {
+    var stock: [Fruits: Int] = [.strawberry: 10, .banana: 10, .pineapple: 10, .kiwi: 10, .mango: 10]
+    private let orderQuantity: [Fruits: Int] = [.strawberry: 80, .banana: 10, .pineapple: 10, .kiwi: 15, .mango: 15]
```
orderQuantity는 부족할때마다 채우기만 하고 양이 줄어들지는 않는건가 보네요? 
그럴수 있다고 생각합니다. 단순 궁금해서 여쭤보아요 ㅎㅎ

yeeton37 replied:
네 부족할 때마다 각 과일 레시피 * 5 만큼씩 채워주기 위해서 저렇게 했습니다..! 제이슨의 말씀대로 과일을 채워주는 메서드를  나중에 구현하기위해서 저 상수는 지워주었습니다..!

ehgud0670 replied:
오오 넵 그럼 각 과일 레시피 * 5 라고 코드로 보이면 좀 더 이해가 될 것 같아요(유연하기도 하고요)
나중에 다시 쓰시면 되겠네요 👍

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
+        }
+    }
+    
+    func fillStock(fruits: [Fruits]) {
```
fillStock은 어제 말씀드린것 처럼 재고가 부족할때 채우는 메소드 말고, 사용자가 재고 플러스 버튼을 눌렀을 때 실행되는 메소드라고 해봐요! 그럼 매개변수가 배열이 아니라 단수 Fruit가 나을 수 있다고 생각합니다 🧐

Hugh-github replied:
`func fillStock(fruit: Fruits, total: Int)`처럼 매개변수 2개를 사용해 fruit을 total 만큼 증가시켜 주는 함수로 변경했습니다.

ehgud0670 replied:
네네 다음에 요구사항에 따라 잘 사용하시면 될 듯합니다 :) 
수정이 필요하면 수정하면 되고요.

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
+        }
+    }
+    
+    func fillStock(fruits: [Fruits]) {
```
fillStock은 어제 말씀드린것 처럼 재고가 부족할때 채우는 메소드 말고, 사용자가 재고 플러스 버튼을 눌렀을때 실행되는 메소드라고 해봐요! 그럼 매개변수가 배열이 아니라 단수 Fruit가 나을 수 있다고 생각합니다 🧐

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
+    private let orderQuantity: [Fruits: Int] = [.strawberry: 80, .banana: 10, .pineapple: 10, .kiwi: 15, .mango: 15]
+    
+    func decreaseStock(menu: Menu, total: Int) {
+        for (fruit, need) in menu.recipe {
```
recipe를 private 으로 접근 제한하고 메소드로 표현해보세요. 메소드의 반환타입은 딕셔너리, 배열 등등 뭐든 좋습니다

yeeton37 replied:
넵 레시피 변수를 private로 접근 제어 해준 뒤에 레시피의 딕셔너리 타입을 반환해주는 메서드를 추가해주었습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
+    private let orderQuantity: [Fruits: Int] = [.strawberry: 80, .banana: 10, .pineapple: 10, .kiwi: 15, .mango: 15]
+    
+    func decreaseStock(menu: Menu, total: Int) {
+        for (fruit, need) in menu.recipe {
```
recipe를 private 으로 접근 제한하고 메소드로 표현해보세요. 메소드의 반환타입은 딕셔너리, 배열 등등 뭐든 좋습니다

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
+        }
+    }
     
+    func unwrapOptional(_ stock: Int?) -> Int {
```
위에서 말씀드린대로 굳이 필요할까 생각이 드네요!

Hugh-github replied:
네......😢

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,24 @@
+        }
+    }
     
+    func unwrapOptional(_ stock: Int?) -> Int {
```
위에서 말씀드린대로 굳이 필요할까 생각이 드네요!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
-// 쥬스 메이커 타입
-struct JuiceMaker {
+struct JuiceMaker: BindingOptional {
+    private var store = FruitStore()
```
FruitStore가 구조체가 아닌 클래스 타입이라면 let으로 선언하는게 더 낫지 않을까요?

yeeton37 replied:
그러네요..! 구조체와 클래스 모두 인스턴스를 선언 할 시 스택의 영역에 들어가게 되는데 클래스는 힙 영역에 값이 들어가고 그 주소값이 스택에 저장되기 때문에  인스턴스를 상수로 선언해도 프로퍼티(변수)의 변경이 가능하겠네요..고쳐주었습니다.!

ehgud0670 replied:
좋습니다 👍

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
-// 쥬스 메이커 타입
-struct JuiceMaker {
+struct JuiceMaker: BindingOptional {
+    private var store = FruitStore()
```
FruitStore가 구조체가 아닌 클래스 타입이라면 let으로 선언하는게 더 낫지 않을까요?

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+struct JuiceMaker: BindingOptional {
+    private var store = FruitStore()
+        
+    private func checkStock(menu: Menu, total: Int) throws {
```
1. 재고를 체크하는 메소드는 재고를 가지고 있는 객체(FruitStore)에서 해줘야 합니다. 왜 재고 체크를 JuiceMaker가 재고 체크가 가능한 상황이 발생했냐면 private으로 변수 선언하지 않아서 이기 때문입니다(결과적으로 결합도가 커졌네요) 전에 말씀드린대로 모든 프로퍼티는 private 으로 접근 제한하고 동작(메소드)을 통해 객체들이 협력하도록 표현하세요.

2. 보통 재고를 체크하는것은 반환타입으로 Bool을 반환하고 에러를 던지지 않습니다. 에러를 사용하는 메소드는 해당 체크 메소드를 사용하는 상위 메소드입니다! 다시 고민해보세요

Hugh-github replied:
회의실에서 얘기한 내용을 바탕으로 
checkStock -> hasFruit으로 변경해 Bool을 반환하고 있습니다.
decreaseStock에서 에러를 던지게 수정했습니다.
decreaseStock을 호출하는 make 함수에서 에러를 처리하고 있습니다.

ehgud0670 replied:
이 부분에 대해선 회의실에서 좀 더 말씀을 드릴게요

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+struct JuiceMaker: BindingOptional {
+    private var store = FruitStore()
+        
+    private func checkStock(menu: Menu, total: Int) throws {
```
1. 재고를 체크하는 메소드는 재고를 가지고 있는 객체(FruitStore)에서 해줘야 합니다. 왜 재고 체크를 JuiceMaker가 재고 체크가 가능한 상황이 발생했냐면 private으로 변수 선언하지 않아서 이기 때문입니다(결과적으로 결합도가 커졌네요) 전에 말씀드린대로 모든 프로퍼티는 private 으로 접근 제한하고 동작(메소드)을 통해 객체들이 협력하도록 표현하세요.

2. 보통 재고를 체크하는것은 반환타입으로 Bool을 반환하고 에러를 던지지 않습니다. 에러를 사용하는 메소드는 해당 체크 메소드를 사용하는 상위 메소드입니다! 다시 고민해보세요

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+            try checkStock(menu: menu, total: total)
+            store.decreaseStock(menu: menu, total: total)
+        } catch {
+            store.fillStock(fruits: classifyKey(from: menu))
```
1. 말씀드린대로 fillStock은 여기서 호출되지 않도록 로직 수정해주세요 
fillStock 는 사용자가 재고 채우는 버튼을 눌렀을때만 유효한 메소드라고 생각하시면 됩니다.

2. 그리고 예외 케이스에 대한 정상 처리를 catch 문에서 하면 논리를 보기가 어렵습니다. 
lackOfStock일때는 말 그대로 재고가 부족하다는 것을 훗날 화면을 만들때 알랏(팝업창)으로 사용자에게 알려주면 되겠습니다 ^^ 

> 예외 케이스에 대한 정상 처리를 catch 문에서 하면 논리를 보기가 어렵습니다. 

=> 이 말은 이해가 안되실수도 있는데 클린코드 7장 오류처리 중 **정상 흐름을 정의하라**에서 나오는 내용입니다.
해당 내용 보시면 훨씬 이해가 잘 되실거에요 
일단 해당 내용 제가 정리한 [링크](https://github.com/BoyScout2021/CleanCode/blob/main/Chapter_07_%EC%98%A4%EB%A5%98_%EC%B2%98%EB%A6%AC/ehgud0670.md#%EC%A0%95%EC%83%81-%ED%9D%90%EB%A6%84%EC%9D%84-%EC%A0%95%EC%9D%98%ED%95%98%EB%9D%BC)를 드리겠습니다! ㅋㅋ

ehgud0670 replied:
https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html

에러핸들링 파트를 보고 한번 더 고민해보세요 🙏
어느 메소드가 에러를 던지면 어울릴지.

ehgud0670 replied:
https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html

에러핸들링 파트를 보고 한번 더 고민해보세요 🙏
어느 메소드가 에러를 던지면 어울릴지.

yeeton37 replied:
오 참고해보겠습니다 감사해요!!!ㅠㅠ

yeeton37 replied:
재고를 채워주는 기능은 나중에 사용자가 버튼을 눌렀을 때 그 개수만큼 채워지도록 코드를 변경해주었습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+            try checkStock(menu: menu, total: total)
+            store.decreaseStock(menu: menu, total: total)
+        } catch {
+            store.fillStock(fruits: classifyKey(from: menu))
```
1. 말씀드린대로 fillStock은 여기서 호출되지 않도록 로직 수정해주세요 
fillStock 는 사용자가 재고 채우는 버튼을 눌렀을때만 유효한 메소드라고 생각하시면 됩니다.

2. 그리고 예외 케이스에 대한 정상 처리를 catch 문에서 하면 논리를 보기가 어렵습니다. 
lackOfStock일때는 말 그대로 재고가 부족하다는 것을 훗날 화면을 만들때 알랏(팝업창)으로 사용자에게 알려주면 되겠습니다 ^^ 

> 예외 케이스에 대한 정상 처리를 catch 문에서 하면 논리를 보기가 어렵습니다. 

=> 이 말은 이해가 안되실수도 있는데 클린코드 7장 오류처리 중 **정상 흐름을 정의하라**에서 나오는 내용입니다.
해당 내용 보시면 훨씬 이해가 잘 되실거에요 
일단 해당 내용 제가 정리한 [링크](https://github.com/BoyScout2021/CleanCode/blob/main/Chapter_07_%EC%98%A4%EB%A5%98_%EC%B2%98%EB%A6%AC/ehgud0670.md#%EC%A0%95%EC%83%81-%ED%9D%90%EB%A6%84%EC%9D%84-%EC%A0%95%EC%9D%98%ED%95%98%EB%9D%BC)를 드리겠습니다! ㅋㅋ

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+        }
+    }
+    
+    private func classifyKey(from juice: Menu) -> [Fruits] {
+        var keys: [Fruits] = []
+        
+        for keyValue in juice.recipe.keys {
+            keys.append(keyValue)
+        }
+        return keys
+    }
     
```
여기 key들도 결국 enum Menu의 receipe의 키들이라 해당 메소드는 Menu로 옮기는게 좋겠습니다.

Hugh-github replied:
fillStock이 자동으로 채워주는 기능이 아니기 때문에 키를 따로 분리하는 함수를 제거해 주었습니다.
기존의 코드에서는 부족한 과일만 확인하기 위해 key를 분리하는 기능을 만들었습니다.

ehgud0670 replied:
애당초 파라미터나 반환타입 모두 고려해보았을때 해당 메소드는 enum Menu에 있는게 어울리다는 겁니다.
classifyKey 네이밍이 의도가 드러나지 않네요 이름이 Fruit가 포함되도록 추후에 만들어보세요

Hugh-github replied:
네 알겠습니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+        }
+    }
+    
+    private func classifyKey(from juice: Menu) -> [Fruits] {
+        var keys: [Fruits] = []
+        
+        for keyValue in juice.recipe.keys {
+            keys.append(keyValue)
+        }
+        return keys
+    }
     
```
여기 key들도 결국 enum Menu의 receipe의 키들이라 해당 메소드는 Menu로 옮기는게 좋겠습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+        return keys
+    }
     
+    func unwrapOptional(_ stock: Int?) -> Int {
+        guard let number = stock else { return 0 }
+        return number
+    }
 }
```
지우시면 좋을 듯 합니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,35 @@
+        return keys
+    }
     
+    func unwrapOptional(_ stock: Int?) -> Int {
+        guard let number = stock else { return 0 }
+        return number
+    }
 }
```
지우시면 좋을 듯 합니다.

Hugh-github replied:
네......😢

ehgud0670 commented on README.md:
```diff
@@ -0,0 +1,131 @@
+
+
+## 배운 것들
+- private을 우선 고려해서 변수나 상수를 선언해보기
```
명심 하십쇼! 무적건 private 변수 선언

yeeton37 replied:
무족권...!!!!!!!

ehgud0670 commented on README.md:
```diff
@@ -0,0 +1,131 @@
+
+
+## 배운 것들
+- private을 우선 고려해서 변수나 상수를 선언해보기
```
명심 하십쇼! 무적건 private 변수 선언

ehgud0670 reviewed (changes requested):
수고하셨습니다. 피드백 반영 부탁드리고 피드백 받고 추가 리뷰 드릴게요!! 😄

ehgud0670 commented:
# 고민했던 것들

==> 이건 오늘 시간이 마땅하지 않아서 추후에 답변드릴게요!

Hugh-github committed:
refactor: BindingOptional 프로토콜 제거

프로토콜 제거 및 프로토콜을 채택한 클래스, 구조체에서 메서드 제거, guard let 바인딩을 이용해 옵셔널 Unwrapping

yeeton37 committed:
refactor: 메서드(classifyKey)를 enum(Menu)안으로 옮겨줌

yeeton37 committed:
feat:enum(Menu)안의 프로퍼티(recipe)를 반환하는 메서드(notifyRecipe) 선언

Hugh-github committed:
refactor: 기존 JuiceMaker에 있던 메서드 checkStock을 FruitStore로 욺김

checkStock -> hasFruit으로 변경
hasFruit의 리턴 타입을 Bool로 변경
FruitStore의 stock 변수의 접근 제어 수준을 private으로 설정

Hugh-github committed:
refactor: fillStock 수정, 에러를 던지는 함수 checkStock 추가, 에러를 make에서 처리

Hugh-github committed:
refactor: 메서드 checkStock 삭제

hasFruit과 중복되는 기능으로 판단되어 삭제
checkStock 삭제로 인해 decreaseStock에서 에러를 던짐
에러 처리는 JuiceMaker의 메서드 make에서 진행

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/Menu.swift:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+    
+    func notifyRecipe() -> [Fruits: Int] {
+        return self.recipe
```
이건 사실 레시피를 그냥 반환하는거랑 차이가 없습니다. 
제가 말씀드리는 건 메소드에 해당 메뉴의 (total:Int)을 파라미터로 받아서 총 필요한 과일 개수들을 반환하는 메소드를 만들어보라는 이야기입니다! (반환타입은 말씀드린대로 배열도 좋고 딕셔너리도 좋습니다)

Hugh-github replied:
`func count(to total: Int, of juice: Self) -> [Fruits: Int]`를 생성해 입력받은 메뉴의 재료와 필요한 개수를 계산해서 반환하는 함수를 생성했습니다.

ehgud0670 replied:
이 부분도 회의실에서 좀 더 말씀을 드릴게요!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,30 @@
 class FruitStore {
+    private var stock: [Fruits: Int] = [.strawberry: 10, .banana: 10, .pineapple: 10, .kiwi: 10, .mango: 10]
+    
+    func hasFruit(menu: Menu, total: Int) -> Bool {
```
hasFruit가 내부에서만 쓰인다면 private 접근제한을 해도 되지 않을까요?

Hugh-github replied:
네 변경했습니다.

yeeton37 replied:
아 그러네요! 변경해주었습니다~

ehgud0670 replied:
추가 첨언하자면 private 으로 둬야하는지 항상 고민하는 습관이 드시면 좋아요!

yeeton37 replied:
오오...명심하겠습니다,,!

yeeton37 committed:
feat: 메서드(notifyRecipe) 제거하고 메서드(count) 선언

- 필요한 과일의 총량을 계산해서 반환해주는 메서드 생성
- 그에 따른 관련 있는 메서드 내부 수정

yeeton37 committed:
refactor: 메서드(hasFruit)의 접근 제어 수준을 private으로 낮춰줌

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/Menu.swift:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+    
+    func count(to total: Int, of juice: Self) -> [Fruits: Int] {
```
파라미터 Juice는 필요없어보여요!

ehgud0670 replied:
그리고 원래 enum 명은 Menu이고 들어온 파라미터는 juice 네요~
변수명이나 타입명이 일관화가 필요해 보입니다.

Hugh-github replied:
네! 명심하겠습니다.

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,40 @@
+                    number >= need
+            else {
+                return false
+            }
```
for 문에 guard 의 else 문에만 코드가 있고
걸리지고 난 뒤에는 코드가 없어서 보기가 조금은 어려워 보이네요
return false 에 대한 코드가 else 가 아닌 guard 에서 걸리진 곳에서 보이는 게 다른 사람들이 보기에 더 잘 보일듯 합니다.

```suggestion
    guard let number = stock[fruit] else { continue }
    if number < need {
        return false 
    }        
```

yeeton37 replied:
훨씬 이해하기 편하네요 감사합니다!

ehgud0670 commented on JuiceMaker/JuiceMaker/Model/Fruits.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by hugh, yeton on 2022/04/26.
+//
+
+enum Fruits {
```
Fruits 는 복수니깐 Fruit가 어떨까요? FruitKind 라던지

yeeton37 replied:
찾아보니 enum 타입명은 보통 단수를 사용하는 군요....!

Hugh-github committed:
refactor: 메서드 hasFruit 가독성 향상을 위한 조건문 변경

Hugh-github committed:
refactor: Fruits -> Fruit으로 변경(파일과 타입이름)

Hugh-github committed:
refactor: 타입 이름(Fruits -> Fruit) 변경으로 인한 코드 수정

Hugh-github committed:
refactor: Menu타입의 메서드 count의 juice 파라미터 제거 및 코드 변경

ehgud0670 reviewed (approved):
이번 리뷰를 통해 배움이 많으셨으면 하네요.
모두 수고하셨습니다!! 👍
